src/hardware/mixer.cpp

```diff
@@ -1052,17 +1050,18 @@ void MixerChannel::Mix(const int frames_requested)
 
 	frames_needed = frames_requested;
 
+	auto freq_add = static_cast<float>(sample_rate_hz) /
+	                static_cast<float>(mixer.sample_rate_hz);
+
 	while (frames_needed > frames_done) {
-		auto frames_remaining = frames_needed - frames_done;
-		frames_remaining *= freq_add;
-		frames_remaining = (frames_remaining >> FreqShift) +
-		                   ((frames_remaining & FreqMask) != 0);
+		auto frames_remaining = iceil(
+		        static_cast<float>(frames_needed - frames_done) + freq_add);
```

Comment:
Again, a little skeptical about this change.  It seems like frames remaining, frames needed, frames done, and frequency should all be integers.  I don't know what the heck the old code was doing with that bit shifting stuff though.  This whole thing just seems kind of weird to me.

Reply:
The old code _did not_ use integers; it used [fixed point arithmetics](https://en.wikipedia.org/wiki/Fixed_point_(mathematics)) to do the same calculations I'm doing now directly with floats.

>  It seems like frames remaining, frames needed, frames done, and frequency should all be integers. 

They are. But `freq_add` wasn't, that was a fixed-point decimal number, with a fractional part, potentially. Now that's a float.